File: edgequake/sdks/typescript/src/resources/workspaces.ts
/**
 * Workspaces resource — workspace management and actions.
 *
 * @module resources/workspaces
 * @see edgequake/crates/edgequake-api/src/handlers/workspaces.rs
 */

import type {
  MetricsHistory,
  UpdateWorkspaceRequest,
  WorkspaceDetail,
  WorkspaceInfo,
  WorkspaceStats,
} from "../types/workspaces.js";
import { Resource } from "./base.js";

export class WorkspacesResource extends Resource {
  /** Get workspace details. */
  async get(workspaceId: string): Promise<WorkspaceDetail> {
    return this._get(`/api/v1/workspaces/${workspaceId}`);
  }

  /** Update workspace settings. */
  async update(
    workspaceId: string,
    request: UpdateWorkspaceRequest,
  ): Promise<WorkspaceInfo> {
    return this._put(`/api/v1/workspaces/${workspaceId}`, request);
  }

  /** Delete a workspace. */
  async delete(workspaceId: string): Promise<void> {
    await this._del(`/api/v1/workspaces/${workspaceId}`);
  }

  /** Get workspace statistics (document count, entity count, etc.). */
  async stats(workspaceId: string): Promise<WorkspaceStats> {
    return this._get(`/api/v1/workspaces/${workspaceId}/stats`);
  }

  /** Get metrics history for a workspace. */
  async metricsHistory(workspaceId: string): Promise<MetricsHistory> {
    return this._get(`/api/v1/workspaces/${workspaceId}/metrics-history`);
  }

  /** Trigger a metrics snapshot. */
  async triggerMetricsSnapshot(workspaceId: string): Promise<void> {
    await this._post(`/api/v1/workspaces/${workspaceId}/metrics-snapshot`);
  }

  /** Rebuild embeddings for a workspace. */
  async rebuildEmbeddings(workspaceId: string): Promise<void> {
    await this._post(`/api/v1/workspaces/${workspaceId}/rebuild-embeddings`);
  }

  /** Rebuild knowledge graph (after LLM model change). */
  async rebuildKnowledgeGraph(workspaceId: string): Promise<void> {
    await this._post(
      `/api/v1/workspaces/${workspaceId}/rebuild-knowledge-graph`,
    );
  }

  /** Reprocess all documents in a workspace. */
  async reprocessDocuments(workspaceId: string): Promise<void> {
    await this._post(`/api/v1/workspaces/${workspaceId}/reprocess-documents`);
  }
}

File: edgequake/sdks/rust/src/resources/workspaces.rs
//! Workspaces resource.

use crate::client::EdgeQuakeClient;
use crate::error::Result;
use crate::types::workspaces::*;

pub struct WorkspacesResource<'a> {
    pub(crate) client: &'a EdgeQuakeClient,
}

impl<'a> WorkspacesResource<'a> {
    /// `GET /api/v1/tenants/{tenant_id}/workspaces`
    pub async fn list(&self, tenant_id: &str) -> Result<Vec<WorkspaceInfo>> {
        self.client
            .get(&format!("/api/v1/tenants/{tenant_id}/workspaces"))
            .await
    }

    /// `POST /api/v1/tenants/{tenant_id}/workspaces`
    pub async fn create(
        &self,
        tenant_id: &str,
        req: &CreateWorkspaceRequest,
    ) -> Result<WorkspaceInfo> {
        self.client
            .post(
                &format!("/api/v1/tenants/{tenant_id}/workspaces"),
                Some(req),
            )
            .await
    }

    /// `GET /api/v1/workspaces/{id}/stats`
    pub async fn stats(&self, workspace_id: &str) -> Result<WorkspaceStats> {
        self.client
            .get(&format!("/api/v1/workspaces/{workspace_id}/stats"))
            .await
    }

    /// `POST /api/v1/workspaces/{id}/rebuild`
    pub async fn rebuild(&self, workspace_id: &str) -> Result<RebuildResponse> {
        self.client
            .post::<(), RebuildResponse>(
                &format!("/api/v1/workspaces/{workspace_id}/rebuild"),
                None,
            )
            .await
    }

    /// `GET /api/v1/workspaces/{id}` — Get workspace by ID.
    pub async fn get(&self, workspace_id: &str) -> Result<WorkspaceInfo> {
        self.client
            .get(&format!("/api/v1/workspaces/{workspace_id}"))
            .await
    }

    /// `PUT /api/v1/workspaces/{id}` — Update workspace.
    pub async fn update(
        &self,
        workspace_id: &str,
        body: &serde_json::Value,
    ) -> Result<WorkspaceInfo> {
        self.client
            .put(
                &format!("/api/v1/workspaces/{workspace_id}"),
                Some(body),
            )
            .await
    }

    /// `DELETE /api/v1/workspaces/{id}` — Delete workspace.
    pub async fn delete(&self, workspace_id: &str) -> Result<()> {
        self.client
            .delete_no_content(&format!("/api/v1/workspaces/{workspace_id}"))
            .await
    }

    /// `GET /api/v1/workspaces/{id}/metrics-history`
    pub async fn metrics_history(&self, workspace_id: &str) -> Result<Vec<serde_json::Value>> {
        self.client
            .get(&format!("/api/v1/workspaces/{workspace_id}/metrics-history"))
            .await
    }

    /// `POST /api/v1/workspaces/{id}/rebuild-embeddings`
    pub async fn rebuild_embeddings(&self, workspace_id: &str) -> Result<serde_json::Value> {
        self.client
            .post::<(), serde_json::Value>(
                &format!("/api/v1/workspaces/{workspace_id}/rebuild-embeddings"),
                None,
            )
            .await
    }

    /// `POST /api/v1/workspaces/{id}/rebuild-knowledge-graph`
    pub async fn rebuild_knowledge_graph(&self, workspace_id: &str) -> Result<serde_json::Value> {
        self.client
            .post::<(), serde_json::Value>(
                &format!("/api/v1/workspaces/{workspace_id}/rebuild-knowledge-graph"),
                None,
            )
            .await
    }

    /// `POST /api/v1/workspaces/{id}/reprocess-documents`
    pub async fn reprocess_documents(&self, workspace_id: &str) -> Result<serde_json::Value> {
        self.client
            .post::<(), serde_json::Value>(
                &format!("/api/v1/workspaces/{workspace_id}/reprocess-documents"),
                None,
            )
            .await
    }
}
